wp-content/civi-extensions/goonjcustom/api/v3/Goonjcustomevent/Cron.php

```diff
@@ -0,0 +1,109 @@
+<?php
+
+/**
+ * @file
+ */
+
+use Civi\Api4\EckEntity;
+use Civi\Api4\MessageTemplate;
+use Civi\Api4\OptionValue;
+
+/**
+ * Civirules.Cron API specification.
+ *
+ * @param array $spec
+ *   Description of fields supported by this API call.
+ *
+ * @return void
+ */
+function _civicrm_api3_goonjcustomevent_cron_spec(&$spec) {
+  // No parameters for the civirules cron.
+}
+
+/**
+ * Civirules.Cron API.
+ *
+ * @param array $params
+ *
+ * @return array API result descriptor
+ *
+ * @throws \CRM_Core_Exception
+ */
+function civicrm_api3_goonjcustomevent_cron($params) {
+  $returnValues = [];
+
+  try {
+    goonjcustomevent_check_and_send_emails_for_camp_end_date();
+  }
+  catch (Exception $e) {
+    error_log('Goonj Cron Job: Error - ' . $e->getMessage());
+  }
+
+  return civicrm_api3_create_success($returnValues, $params, 'Goonjcustomevent', 'cron');
+}
+
+/**
+ * Function to check camp end date and send emails.
+ */
+function goonjcustomevent_check_and_send_emails_for_camp_end_date() {
+  $optionValues = OptionValue::get(FALSE)
+    ->addWhere('option_group_id:label', '=', 'ECK Subtypes')
+    ->addWhere('label', '=', 'Collection Camp')
+    ->execute();
+
+  $collectionCampSubtype = $optionValues->first()['value'];
+  try {
+    $collectionCamps = EckEntity::get('Collection_Camp', TRUE)
+      ->addSelect('Logistics_Coordination.Camp_to_be_attended_by', 'Collection_Camp_Intent_Details.End_Date')
+      ->addWhere('subtype', '=', $collectionCampSubtype)
+      ->execute();
+
+    // Get today's date in the same format as the End_Date field.
+    $today = new DateTime();
+    $todayFormatted = $today->format('Y-m-d');
+
+    foreach ($collectionCamps as $camp) {
+      if (empty($camp['Logistics_Coordination.Camp_to_be_attended_by'])) {
+        continue;
+      }
+      $recipientId = $camp['Logistics_Coordination.Camp_to_be_attended_by'];
+      $endDate = new DateTime($camp['Collection_Camp_Intent_Details.End_Date']);
+      $endDateFormatted = $endDate->format('Y-m-d');
+
+      // Only send the email if the end date is exactly today.
+      if ($endDateFormatted === $todayFormatted) {
```

Comment:
I have avoided using direct date comparisons in the query because the time component plays a crucial role. Instead, we fetch the end dates first and then convert them to the same format as today's date before performing the comparison.

```
      $endDate = new DateTime($camp['Collection_Camp_Intent_Details.End_Date']);
      $endDateFormatted = $endDate->format('Y-m-d');
```

Reply:
> the time component plays a crucial role

Can you explain this a bit more? Specific to our case.

Reply:
It's not even an optional thing – considering the performance impact. Your current query will fetch ALL the collection camps present in the system and then loop over them. Now as the system grows, the number of collection camps will continue to grow and your function will keep on getting slower and slower.

But if you add an additional where clause for end date, no matter how many collection camps there are in the database, you will only get a limited number of collection camps that are active in that case.

Think:
- how many collection camps will be there in the database today
- how many collection camps will be there in the database after an year

Now think

- how many ACTIVE collection camps will be there in the database today
- how many ACTIVE collection camps will be there in the database after a year